pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTWhen.java

```diff
@@ -25,29 +25,73 @@
 
 package org.jenkinsci.plugins.pipeline.modeldefinition.ast;
 
+import net.sf.json.JSONArray;
 import net.sf.json.JSONObject;
 import org.jenkinsci.plugins.pipeline.modeldefinition.validator.ModelValidator;
 
+import java.util.ArrayList;
+import java.util.List;
+
 /**
- * Represents a block for when/if a {@link ModelASTStage} will be executed or not.
+ * If {@link ModelASTStage} will be executed or not.
  */
-public class ModelASTWhen extends AbstractModelASTCodeBlock {
+public class ModelASTWhen extends ModelASTElement {
+
+    private List<ModelASTStep> conditions = new ArrayList<>();
+
     public ModelASTWhen(Object sourceLocation) {
-        super(sourceLocation, "when");
+        super(sourceLocation);
+    }
+
+    public List<ModelASTStep> getConditions() {
+        return conditions;
+    }
+
+    public void setConditions(List<ModelASTStep> conditions) {
+        this.conditions = conditions;
+    }
+
+    @Override
+    public Object toJSON() {
+        final JSONArray a = new JSONArray();
+        for (ModelASTStep c: conditions) {
+            a.add(c.toJSON());
+        }
+        return new JSONObject().accumulate("conditions", a);
     }
 
     @Override
-    public JSONObject toJSON() {
-        JSONObject o = new JSONObject();
-        if (getArgs() != null) {
-            o.accumulate("arguments", getArgs().toJSON());
+    public String toGroovy() {
+        StringBuilder result = new StringBuilder("when {\n");
+        for (ModelASTStep c: conditions) {
+            result.append(c.toGroovy()).append("\n");
         }
-        return o;
+        result.append("}\n");
+        return result.toString();
     }
 
     @Override
-    public void validate(ModelValidator validator) {
-        super.validate(validator);
+    public void removeSourceLocation() {
+        super.removeSourceLocation();
```

Comment:
To get the source location information (i.e., the `ASTNode` or `JSONObject`) out so that we can serialize as a graph - e.g., https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy#L53 calling https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy#L192-L201